src/stan_math_backend/Transform_Mir.ml

```diff
@@ -64,49 +64,72 @@ let rec switch_expr_to_opencl available_cl_vars (Expr.Fixed.({pattern; _}) as e)
         pattern=
           Expr.Fixed.Pattern.map (switch_expr_to_opencl available_cl_vars) x }
 
+let rec base_type = function
+  | SizedType.SArray (t, _) -> base_type t
+  | SVector _ | SRowVector _ | SMatrix _ -> UnsizedType.UReal
+  | x -> SizedType.to_unsized x
+
 let pos = "pos__"
 
 let data_read smeta (decl_id, st) =
+  let unsized = SizedType.to_unsized st in
+  let scalar = base_type st in
+  let flat_type = UnsizedType.UArray scalar in
   let decl_var =
     { Expr.Fixed.pattern= Var decl_id
-    ; meta=
-        Expr.Typed.Meta.
-          {loc= smeta; type_= SizedType.to_unsized st; adlevel= DataOnly} }
+    ; meta= Expr.Typed.Meta.{loc= smeta; type_= unsized; adlevel= DataOnly} }
   in
   let swrap stmt = {Stmt.Fixed.pattern= stmt; meta= smeta} in
-  let bodyfn var =
-    let pos_var = {Expr.Fixed.pattern= Var pos; meta= Expr.Typed.Meta.empty} in
-    let readfnapp var =
-      let f =
-        Expr.Helpers.internal_funapp FnReadData
-          [{var with pattern= Lit (Str, decl_id)}]
-          var.meta
+  let pos_var = {Expr.Fixed.pattern= Var pos; meta= Expr.Typed.Meta.empty} in
+  let readfnapp var =
+    Expr.Helpers.internal_funapp FnReadData
+      [{var with pattern= Lit (Str, decl_id)}]
+      Expr.Typed.Meta.{var.meta with type_= flat_type}
+  in
+  match unsized with
+  | UInt | UReal ->
+      [ Assignment
+          ( (decl_id, unsized, [])
+          , { Expr.Fixed.pattern=
+                Indexed (readfnapp decl_var, [Single Expr.Helpers.loop_bottom])
+            ; meta= {decl_var.meta with type_= unsized} } )
+        |> swrap ]
+  | UArray UInt | UArray UReal ->
+      [Assignment ((decl_id, flat_type, []), readfnapp decl_var) |> swrap]
+  | _ ->
```

Comment:
I think it would be good to explicitly list the types we're matching on here because in the future we may add new types and we want the compiler to warn us about places where we'll need to make decisions.

Reply:
Agreed. I guess UFun and UMathLibraryFunction shouldn't be allowed?

Reply:
probably not, yeah